docs/proposals/20200827-history.md

```diff
@@ -0,0 +1,175 @@
+
+#### When will history be stored
+
+Everytime `kconnect use` is invoked and a kubeconfig file is created a history item will be created.
+
+#### Data stored
+
+The data stored in the history file will be an entry for each successful `kconnect use` invocation. Each entry will store the following:
+
+* Disovery provider used (i.e. eks)
+* Identity provider used (i.e. saml)
+* Date/time (ISO 8601 format) of when the entry was added
+* Unique discovery provider specific identifer for that cluster. For example, with the EKS provider we might stote the AWs account ID and EKS cluster name. This identifier must be unique and stable.
+* Alias - optional name that can be used
+* Flags and flag values - except flags deemded sensitive
+* Location of the kubeconfig file that was written to
+
+As an example (if we where using YAML as the format):
+
+```yaml
+-
+provider: eks
+identity: saml
+id: MTIzNDU2Nzg5L215Y2x1c3RlcjE  # base 64 url safe encoding of 123456789/mycluster1
+datetime: 2020-08-27T16:06:08.180856
+flags:
+    username: bob@someone.com
+    idp-provider: GoogleApps
+    idp-endpoint: https://accounts.google.com/o/saml2/initsso?idpid=
+    profile: g-saml
+    role: arn:aws:iam::123456789:role/TestAccess
+config: ~/.kube/config
+alias: dev-eks-cluster
+```
+
+### User Experience
+
+#### Showing the history
+
+First showing help for the `ls` command
+
+```bash
+kconnect ls -h
+
+shows the history of the connections you have made using kconnect
+
+Usage:
+
+kconnect ls [FLAGS] e.g ./kconnect ls --provider eks
+
+flags:
+
+-p  --provider         [optional]    used to filter history items based on a specific cluster provider (e.g. eks)
+-i  --identity         [optional]    used to filter history items based on a specific identity provider (e.g. saml)
+    --id               [optional]    used to filter history items for a cluster with the specified id
+-a  --alias            [optional]    used to filter history items for a cluster with a specific alias. It supports wildcards.
+-f  --flag             [optional]    used to filter history items based on a flag and its value
+-k  --kubeconfig       [optional]    used to filter history items that where written to the specified kubeconfig
+-o  --output           [optional]    used to specify the output format of the history. Defaults to yaml. Options are [yaml,json,table]
+
+Example usage:
+
+    kconnect ls --flag username=bob@somewhere.com
+
+```
+
+When using the command to filter connections made to EKS you might do the following:
+
+```bash
+kconnect ls --provider eks -o table
+
+| ID                        | Providers | Identity | Alias       | Config     | Flags                      |
+|===========================|===========|==========|=============|============|============================|
+|MTIzNDU2Nzg5L215Y2x1c3RlcjE| EKS       | SAML     | dev-eks     | (default)  | username=bob@somewhere.com |
```

Comment:
Should we decode this base64 value to display a user friendly value?

Reply:
I have updated the proposal to include 2 different identifiers:

- **id** which is a unique history identifier
- **providerId** which is generated by the providers and is only meant to be understood by the provider

So to answer the queery i don't think we should decode it.